eclair-core/src/main/scala/fr/acinq/eclair/channel/fund/InteractiveTxBuilder.scala

```diff
@@ -105,12 +105,28 @@ object InteractiveTxBuilder {
                                  isInitiator: Boolean,
                                  localAmount: Satoshi,
                                  remoteAmount: Satoshi,
+                                 commonInput_opt: Option[InputInfo],
                                  fundingPubkeyScript: ByteVector,
                                  lockTime: Long,
                                  dustLimit: Satoshi,
                                  targetFeerate: FeeratePerKw,
-                                 requireConfirmedInputs: RequireConfirmedInputs) {
-    val fundingAmount: Satoshi = localAmount + remoteAmount
+                                 requireConfirmedInputs: RequireConfirmedInputs,
+                                 localSpliceOut_opt: Option[TxOut] = None,
+                                 remoteSpliceOut_opt: Option[TxOut] = None) {
+    val previousAmount: Satoshi = commonInput_opt.map(_.txOut.amount).getOrElse(0 sat)
+
+    val localSpliceOutAmount: Satoshi = localSpliceOut_opt.map(_.amount).getOrElse(0 sat)
+    val remoteSpliceOutAmount: Satoshi = remoteSpliceOut_opt.map(_.amount).getOrElse(0 sat)
+    // splice-out are special, because there is potentially no additional funding, so the mining fees must be taken from the previous funding amount
+    // if there is one splice-out, the side that does it pays; if both sides do a splice-out, the initiator pays
+    // TODO: weight is hardcoded and is the same whether there are one or two splices, it should be fully computed using the actual weight of each splice output
```

Comment:
I don't think this is the right model, why diverge from the `interactive-tx` specification?
I think it should work this way:

- we don't need to know `remoteSpliceOut_opt` beforehand and we don't need the `SpliceOutOutputTlv` field, they should be removed
- the only thing we need to know is how much our peer will contribute to the new funding output, which is already provided by the `SharedOutputContributionTlv` (if that is less than their current balance, we can infer they're doing a splice-out, but we don't even really need to care)
- we split on-chain fees according to the `interactive-tx` spec: each peer pays for their own inputs/outputs, and the initiator of the splice pays for the weight of the current funding input, the new funding output, and the common transaction fields

The flow when Alice and Bob are both splicing out would then look like:

- we currently have a channel where balanceA = 50 000 sat and balanceB = 100 000 sat
- alice initiates a splice-out and says she'll contribute 25 000 sats to the new funding output
- bob acks that operation and says that he'll contribute 50 000 sats to the new funding output
- they independently compute the on-chain fees they will need to pay for their part of the transaction, and thus compute the amount they can put in their splice output, for example outputAmountA = 23 000 sat (alice pays 2 000 sats on-chain fees) and outputAmountB = 49 000 sats (bob pays 1 000 sats on-chain fees)
- when they exchange `tx_complete`, each side verifies that the other side correctly paid the fees

This doesn't need adding new fields and I believe better matches the `interactive-tx` flow. It is also more general: you don't need to know beforehand what your peer will do, you just let them add inputs/outputs using `interactive-tx` and simply verify that fees are correctly paid when the tx is complete.

If we want the peer to pay for the service we're providing, that just needs to be done using `push_amount` independently of the on-chain part.

Conceptually, the difference with what you're currently doing is that you're taking the fees from the new funding output, whereas I'm suggesting taking them out of each peer's outputs. I think the reason you did that is because from a UX's point of view, users choose the amount they want to splice-out, not the amount they want for the new channel capacity. But each can be compute deterministically from the other one:

- the user has current balance M sats and wants to splice-out N sats at a given feerate
- they're immediately able to compute all the transaction parts that they'll pay, so they know exactly how much on-chain fees they need to pay (F sats)
- they can then send a splice message where their local contribution is (M - N - F) sats

Reply:
Just to make sure I understand, in your example alice is splicing out 25 000 sat, bob is splicing out 50 000 sat, and none are splicing-in anything?

> we currently have a channel where balanceA = 50 000 sat and balanceB = 100 000 sat
> alice initiates a splice-out and says she'll contribute 25 000 sats to the new funding output
> bob acks that operation and says that he'll contribute 50 000 sats to the new funding output

This implies that `fundingContribution` should _include_ existing balance, while I'm treating it as _additional_ funding. The reason I went this way is because the balance is in `msat`, whereas funding is in `sat`. I saw this as a reason to not mix the two. 

> This doesn't need adding new fields and I believe better matches the interactive-tx flow.

AFAICT the existing fee considerations happen in `InteractiveTxFunder`. But a splice-out can (and will most of the time) happen without a splice-in, so we are bypassing `InteractiveTxFunder` entirely?

Reply:
> This implies that fundingContribution should include existing balance, while I'm treating it as additional funding. The reason I went this way is because the balance is in msat, whereas funding is in sat. I saw this as a reason to not mix the two.

But then you need to allow negative funding contributions for splicing out, which is...ugh!
I think it's conceptually simpler if `fundingContribution` represents the total funding contribution in the new channel.
I agree that means we may need to truncate balances to satoshis, but that's probably ok, isn't it? It just means we'll add 1 sat more to on-chain fees? I'm curious to see what the current spec draft does here, I'll have a look.

> AFAICT the existing fee considerations happen in InteractiveTxFunder.

I meant the fee verification that is done in `validateTx` in `InteractiveTxBuilder`. It does verify everything, except witness weight, but that has no impact for splice-out (since in that case we're not adding any new input and know the witness weight for spending the current funding output).